### PR TITLE
Feature:: Preprocessing models

### PR DESCRIPTION
**Pull request by AHReccese**
#### Reference Issues/PRs
#88 

#### What does this implement/fix? Explain your changes.
In this PR following changes are applied:
##### Following preprocessing classes are implemented and tested. In front of each module, I wrote the problem we faced when adding it to Pymilo.

+ [x] FunctionTransformer
+ [x] KBinsDiscretizer ( :no_entry:  Shape miss-match `array([[..],...,[...]])` and `array([array([..]),...,array([...])])`
+ [x] KernelCenterer
+ [x] MultiLabelBinarizer
+ [x] MaxAbsScaler
+ [x] Normalizer
+ [x] OrdinalEncoder ( :warning:  when saving the JSON file one of `encoded_missing_value` field could be `NaN` which I think can make JSON file unparsable)
+ [x] PolynomialFeatures
+ [x] PowerTransformer ( :no_entry: Dependent on the `StandardScaler` transportation - transportation loop)
+ [x] QuantileTransformer
+ [x] RobustScaler
+ [x] SplineTransformer ( :no_entry: Dependent on the `scipy.interpolate.BSpline`)
+ [x] TargetEncoder ( :no_entry: `sklearn.__version__ > 1.3`)

##### Tests for the following preprocessing classes are tested:
+ [x] StandardScaler
+ [x] OneHotEncoder
+ [x] LabelBinarizer
+ [x] LabelEncoder

#### Any other comments?

### CODE REVIEW

**Comment by AHReccese** [comment]
@sadrasabouri 
Dear Sadra, please review my updates.

**Comment by AHReccese** [comment]
NaN-type support is also added.
Also please listen to voices in telegram, I've explained what I've done to handle these cases:
- KBinsDiscretizer
- PowerTransformer
- SplineTransformer
- TargetEncoder
- OrdinalEncoder

**Comment by AHReccese** [comment]
We will address the coverage issue in a separate PR.
I will enhance code coverage to over 90%.

File: pymilo/transporters/preprocessing_transporter.py
Comment (sadrasabouri) [inline pymilo/transporters/preprocessing_transporter.py:116]: We can also use the `value` in following `if` statement conditions. If you are not using `value`, you can simply iterate over `data`.

**Review by sadrasabouri — CHANGES_REQUESTED** [review]
Thanks Amirhossein. I left minor comments.

File: CHANGELOG.md
Comment (AHReccese) [reply CHANGELOG.md:15]: What's the former format of this line?

File: SUPPORTED_MODELS.md
Comment (AHReccese) [reply SUPPORTED_MODELS.md:688]: Sure Sure

File: pymilo/utils/util.py
Comment (AHReccese) [reply pymilo/utils/util.py:137]: you are right.

File: pymilo/transporters/general_data_structure_transporter.py
Comment (AHReccese) [reply pymilo/transporters/general_data_structure_transporter.py:469]: I will update this part, you're right.

File: pymilo/transporters/preprocessing_transporter.py
Comment (AHReccese) [reply pymilo/transporters/preprocessing_transporter.py:8]: `scipy` is getting import by scikit implicitly

File: pymilo/transporters/preprocessing_transporter.py
Comment (AHReccese) [reply pymilo/transporters/preprocessing_transporter.py:116]: good point, sure.

**Comment by AHReccese** [comment]
Dear Sadra, I've applied your suggestion, thank you so much for your detailed review. I really appreciate it.
Please re-review again.

File: CHANGELOG.md
Comment (sadrasabouri) [reply CHANGELOG.md:15]:
```
`util.py` in chains
```

Comment (AHReccese) [reply CHANGELOG.md:15]: what's the difference? 
@sadrasabouri

Comment (sadrasabouri) [reply CHANGELOG.md:15]: You changed it to :
```
`util.py` in `chains`
```

Comment (AHReccese) [reply CHANGELOG.md:15]: oooh right, could you please apply these changes by yourself? you are way more accurate than I am :)
I completely agree.
@sadrasabouri

Comment (sadrasabouri) [reply CHANGELOG.md:15]: Sure

**Review by sadrasabouri — APPROVED** [review]
LGTM now,
@sepandhaghighi can you please review it?

**Comment by sepandhaghighi** [comment]
> LGTM now, @sepandhaghighi can you please review it?

I will review this pull request within the next 48 hours. 💯

File: pymilo/transporters/preprocessing_transporter.py
Comment (sepandhaghighi) [reply pymilo/transporters/preprocessing_transporter.py:8]: We need to add `scipy` to both `requirements.txt` and `dev-requirements.txt` because there is no guarantee that `scipy` will remain an essential requirement for `scikit-learn` in the future.

File: CHANGELOG.md
Comment (sepandhaghighi) [inline CHANGELOG.md:24]: remove this line

File: SUPPORTED_MODELS.md
Comment (sepandhaghighi) [inline SUPPORTED_MODELS.md:1]: `Last Update: 2024-05-30`

File: pymilo/utils/util.py
Comment (sepandhaghighi) [inline pymilo/utils/util.py:128]: I recommend using `list1` and `list2` instead of `a` and `b`.

**Review by sepandhaghighi — CHANGES_REQUESTED** [review]
@AHReccese, @sadrasabouri Thanks for your efforts! 💯 
Please review my comments. 🔥

File: pymilo/utils/util.py
Comment (sadrasabouri) [reply pymilo/utils/util.py]: fixed in [a600f62](https://github.com/openscilab/pymilo/pull/104/commits/a600f6272c3a50422d6ef57f75d124ba6c63694a).

File: CHANGELOG.md
Comment (sadrasabouri) [reply CHANGELOG.md]: fixed in [3d6eacb](https://github.com/openscilab/pymilo/pull/104/commits/3d6eacb5a617611e808f9fb2aa07f9e3c8d8c701).

File: pymilo/transporters/preprocessing_transporter.py
Comment (sadrasabouri) [reply pymilo/transporters/preprocessing_transporter.py:8]: fixed [4758dcf](https://github.com/openscilab/pymilo/pull/104/commits/4758dcfe873716c9f34417714750a5905a7aea73).

**Comment by AHReccese** [comment]
@sadrasabouri 
Thank you Sadra for applying Sepand's comments, I went through them one by one and it is great!
@sepandhaghighi waiting for your final review sir.

**Review by sepandhaghighi — APPROVED** [review]
LGTM 🔥

**Pull request merged**
